File: source/98.验证二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=98 lang=cpp
 *
 * [98] 验证二叉搜索树
 */
#include "cpp_includes.h"
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long pre = LONG_MIN;

    bool isValidBST(TreeNode* root) {
        if (!root) return true;
        
        if (!isValidBST(root->left)) return false;

        if (root->val <= pre) return false;
        pre = root->val;

        return isValidBST(root->right);
    }
};
// @lc code=end

File: 98.验证二叉搜索树.py
#
# @lc app=leetcode.cn id=98 lang=python3
#
# [98] 验证二叉搜索树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.pre = None

    def isValidBST(self, root: TreeNode) -> bool:
        if root is None: return True

        if self.isValidBST(root.left) == False: return False

        if self.pre is not None:
            if root.val <= self.pre.val: return False
        self.pre = root

        return self.isValidBST(root.right)

# @lc code=end

